lib/developer_portal/app/controllers/developer_portal/admin/account/invoices_controller.rb

```diff
@@ -1,34 +1,88 @@
+# frozen_string_literal: true
+
 class DeveloperPortal::Admin::Account::InvoicesController < ::DeveloperPortal::BaseController
+  class PaymentGatewayNotSupported < StandardError; end
+
+  rescue_from PaymentGatewayNotSupported, with: :handle_not_found
 
   helper Finance::InvoicesHelper
   helper Accounts::InvoicesHelper
 
   before_action :authorize_finance
   before_action :find_provider
+  before_action :authorize_payment_gateway, only: %i[payment payment_succeeded]
+  before_action :find_invoice, only: %i[show payment payment_succeeded]
+
   activate_menu :account, :invoices
 
   liquify prefix: 'invoices'
 
   def index
-    collection = current_account.invoices.visible_for_buyer.page(params[:page])
-
+    collection = accessible_invoices.page(params[:page])
     invoices = Liquid::Drops::Invoice.wrap(collection)
     pagination = Liquid::Drops::Pagination.new(collection, self)
     assign_drops invoices: invoices, pagination: pagination
   end
 
   def show
-    invoice = current_account.invoices.visible_for_buyer.find(params[:id])
     assign_drops invoice: Liquid::Drops::Invoice.wrap(invoice)
   end
 
+  def payment
+    @payment_intent = @invoice.payment_intents.latest_pending.first!
```

Comment:
:warning: Objects leaked in the liquid page should be wrapped into a `Liquid::Drop` otherwise man can access to any ORM object in the database.
This is an issue because a provider can just access master and change whatever there ...

Reply:
Second thoughts, payment is an ERB template so it should be fine.